execution-plane/arcon/arcon_state/src/faster/aggregator_ops.rs

```diff
@@ -0,0 +1,106 @@
+use crate::{
+    error::*, serialization::protobuf, Aggregator, AggregatorOps, AggregatorState, Faster, Handle,
+    Metakey,
+};
+
+pub(crate) const ACCUMULATOR_MARKER: u8 = 0xAC;
+pub(crate) const VALUE_MARKER: u8 = 0x00;
+
+impl AggregatorOps for Faster {
+    fn aggregator_clear<A: Aggregator, IK: Metakey, N: Metakey>(
+        &mut self,
+        handle: &Handle<AggregatorState<A>, IK, N>,
+    ) -> Result<()> {
+        let key = handle.serialize_id_and_metakeys()?;
+        self.remove(&key)?;
+        Ok(())
+    }
+
+    fn aggregator_get<A: Aggregator, IK: Metakey, N: Metakey>(
+        &self,
+        handle: &Handle<AggregatorState<A>, IK, N>,
+    ) -> Result<<A as Aggregator>::Result> {
+        let key = handle.serialize_id_and_metakeys()?;
+
+        if let Some(serialized) = self.get_agg(&key)? {
+            assert_eq!(serialized[0], ACCUMULATOR_MARKER);
+            let serialized = &serialized[1..];
+
+            let current_accumulator = protobuf::deserialize(serialized)?;
+            Ok(handle
+                .extra_data
+                .accumulator_into_result(current_accumulator))
+        } else {
+            Ok(handle
+                .extra_data
+                .accumulator_into_result(handle.extra_data.create_accumulator()))
+        }
+    }
+
+    fn aggregator_aggregate<A: Aggregator, IK: Metakey, N: Metakey>(
+        &mut self,
+        handle: &Handle<AggregatorState<A>, IK, N>,
+        value: <A as Aggregator>::Input,
+    ) -> Result<()> {
+        let key = handle.serialize_id_and_metakeys()?;
+        let mut serialized = Vec::with_capacity(1 + protobuf::size_hint(&value).unwrap_or(0));
+        serialized.push(VALUE_MARKER);
+        protobuf::serialize_into(&mut serialized, &value)?;
+
+        // See the make_aggregate_fn function in this module. Its result is set as the
+        // merging operator for this state.
+        self.aggregate(&key, serialized, handle.id)
+    }
+}
+
+pub fn make_aggregate_fn<A>(aggregator: A) -> Box<dyn Fn(&[u8], &[u8]) -> Vec<u8> + Send>
```

Comment:
Quite many  ``.except()`` cases in this function. Perhaps the return should be a Result<>?

Reply:
It's not very easy to do, because the expects are inside the closure that is returned. Hmm... I'd have to add an error-carrying variant to the types that implement `FasterRmw`, where the closures ultimately get used. That's probably a better solution, because panicking across ffi boundaries is UB as far as I know.

Reply:
Fixed in 8cf8d39